Fixes #28595 - move API actions to an helper action

Ron-Lavi opened the pull request:
instead of writing every API action for every component,
it will be better to consume it from the API Actions file,
It will also be easier to maintain in one place.


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

sharvit commented on webpack/assets/javascripts/react_app/components/Bookmarks/__tests__/__snapshots__/integration.test.js.snap:
```diff
@@ -5,7 +5,13 @@ Object {
   "action": Array [
     Array [
       Object {
+        "actionTypes": Object {
```
Isn't it odd this change adds `actionTypes` and `params` to the snap?

sharvit commented on webpack/assets/javascripts/react_app/redux/API/APIActions.js:
```diff
@@ -0,0 +1,22 @@
+  payload,
+  params,
+  actionTypes,
+});
```
Could it be more simple?
```js
export const get = options => ({ type: API_OPERATIONS.GET, ...options });
```

sharvit commented on webpack/assets/javascripts/react_app/redux/API/APIActions.js:
```diff
@@ -0,0 +1,22 @@
+
+export default {
+  get,
+};
```
We don't have to mix named with default imports.
It can be removed and consumed as:
```js
import * as ApiActions from './ApiActions`;
```

And in the `API/index.js` it can be:
```js
export * as ApiActions from './APIActions`;
```

sharvit reviewed (commented):
Thanks @laviro

Ron-Lavi commented:
Thanks @sharvit, updated with your suggestions !

sharvit commented on webpack/assets/javascripts/react_app/redux/API/APIActions.js:
```diff
@@ -0,0 +1,12 @@
+ * @param { String } options.key the unique key of the API request, will be used in the selector too.
+ * @param { String } options.url the url for the API request.
+ * @param { Object } options.params the API get request params.
+ * @param { Object } options.payload the API payload which will be passed also to the reducer.
```
I'm confused, didn't see the `params` before. Didn't we decide to use the `payload` as the `params`?

laviro replied:
Yes, sorry about that.. it was written before the discussion in #7301

Ron-Lavi replied:
These are the API request params, payload will be passed to the reducer if needed

sharvit replied:
I get it now, sorry for the confusion.
In this case, can the `url` and the `params` move to the `payload`?

Then we will keep in the state as **the payload that caused this response**.
The payload can be used as a hack to pass more data to the reducer but I don't think it's a valid use-case.

Ron-Lavi replied:
currently we have a **technical debt** in many components that use the payload to pass some `id` or `controller` to set substates. soon after we'll replace all of those reducers by the main API reducer #7222 we could drop the payload from the API actions in my opinion - and we will move the `key`, `url` and `params` into the payload as you mentioned.

I think moving all of the current API actions into an action creator would really make it easy as here we only receive those as function arguments, but we can play with the structure and manage it in one place.

sharvit commented on webpack/stories/docs/APIMiddleware.md:
```diff
@@ -1,48 +1,59 @@
+    key: MY_SPECIAL_KEY, // this will be used later to identify your API call, so keep it unique.
+    url: url,
+    params: {} // some params you will want to pass to the API request.
+    payload: { data } // some data to pass to your reducer.
```
It's confusing, it gives me the impression that I must have a `data` key in my payload.

sharvit reviewed (commented):
Thanks @laviro

MariaAga commented on webpack/stories/docs/APIMiddleware.md:
```diff
@@ -1,48 +1,59 @@
+const someAction = url =>
+  get({
+    key: MY_SPECIAL_KEY, // this will be used later to identify your API call, so keep it unique.
+    url: url,
```
can be just `url,` to avoid eslint error when copying

Ron-Lavi replied:
I documented it this way so it would be more readable for non js develooers too, but I agree with you that eslint will raise an error on it.

MariaAga commented on webpack/stories/docs/APIMiddleware.md:
```diff
@@ -1,48 +1,59 @@
+  get({
+    key: MY_SPECIAL_KEY, // this will be used later to identify your API call, so keep it unique.
+    url: url,
+    params: {} // some params you will want to pass to the API request.
```
Currently the middleware uses ` payload.params || {}` and not params. is it going to change?

Ron-Lavi replied:
As discussed in https://github.com/theforeman/foreman/pull/7308#discussion_r362395295 after the technical debt of this components will be solved, I would like everything to go inside the payload, and remove the current option to pass your own payload, although when using the action creator the consumer shouldn't care much how it is implemented.

MariaAga replied:
I agree that the user should not care about the implementation, but should it change in this PR?
or should the `get` function do
``` JS 
payload: {
...payload, params
}
```
?

Ron-Lavi replied:
Yes I was thinking about something like that.. will update soon this PR

ronlavi2412 committed:
Fixes #28595 - move API actions to helper action

instead of writing every API action for every component,
it will be better to consume it from the API Actions file,
it will also be easier to maintain in one place.
refactored the action structure -
moved everything into its payload instead of using many keys.

Ron-Lavi commented:
thanks @sharvit @MariaAga, I didn't want this PR to grow much
but I agree we should fix the action structure too, instead of many keys, I used the convention of `type` and `payload`.

I moved all of the keys to live inside the `payload`.
as I mentioned, currently many components are passing a `payload` to adjust their own state,
soon as #7222 will get merge, we won't need that nested `payload`, and we will delete it from the action.

sharvit commented on webpack/stories/docs/APIMiddleware.md:
```diff
@@ -1,48 +1,62 @@
+    url,
+    params: {}, // some params you will want to pass to the API request.
+    headers: {}, // some headers you will want to pass to the API request.
+    payload: {}, // some payload data to pass to your reducer.
```
Can we remove the `payload` from here and keep it as an internal-hack?

sharvit reviewed (commented):
Thanks @laviro

sharvit reviewed (approved):
Thanks @laviro and @MariaAga , LGTM :+1: 

@MariaAga, waiting for you to give you ack before merging.
